File: 剑指Offer/34.二叉树中和为某一值的路径.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    vector<vector<int>> res;
    void dfs(vector<int>& singlePath, TreeNode* root, int curSum, int sum) {
        if (root == NULL) return;

        curSum += root->val;
        singlePath.push_back(root->val);
        if (root->left == NULL && root->right == NULL && curSum == sum) {
            res.emplace_back(singlePath);
        }
        else {
            dfs(singlePath, root->left, curSum, sum);
            dfs(singlePath, root->right, curSum, sum);
        }
        singlePath.pop_back();
    }

public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<int> singlePath;
        dfs(singlePath, root, 0, sum);
        return res;
    }
};

File: 剑指Offer/34.二叉树中和为某一值的路径.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        def dfs(singlePath: List[int], root: TreeNode, curSum: int, sum: int) -> None:
            if root == None:
                return None

            curSum += root.val
            singlePath.append(root.val)
            if root.left == None and root.right == None and curSum == sum:
                res.append(singlePath[:])
            else:
                dfs(singlePath, root.left, curSum, sum)
                dfs(singlePath, root.right, curSum, sum)
            
            singlePath.pop()

        res = []
        singlePath = []
        dfs(singlePath, root, 0, sum)
        return res